tests/5.0/loop/test_loop_collapse_device.F90

```diff
@@ -0,0 +1,122 @@
+!===--- test_loop_collapse_device.F90 -------------------------------------===//
+!
+! OpenMP API Version 5.0 Nov 2018
+!
+! This test uses the collapse clause with the loop directive and tests that
+! for loops out of the scope of the collapsed loops are not parallelized.
+! This test tests using one and two collapsed loops. This test checks these
+! features in an offloading (target) context.
+!
+!//===----------------------------------------------------------------------===//
+
+#include "ompvv.F90"
+
+! Array Size of 128 uses 16MB target memory and
+! scales n^3 in test_collapse2()
+#define N 128
+
+PROGRAM test_loop_collapse_device
+  USE iso_fortran_env
+  USE ompvv_lib
+  USE omp_lib
+  implicit none
+
+  OMPVV_TEST_OFFLOADING
+
+  OMPVV_TEST_VERBOSE(test_collapse1() .ne. 0)
+  OMPVV_TEST_VERBOSE(test_collapse2() .ne. 0)
+
+  OMPVV_REPORT_AND_RETURN()
+
+CONTAINS
+  INTEGER FUNCTION test_collapse1()
+    INTEGER,DIMENSION(N,N):: a
+    INTEGER,DIMENSION(N+1,N):: b
+    INTEGER:: errors, x, y, temp_total
+
+    errors = 0
+
+    ! a and b array initialization
+    DO x = 1, N
+       b(1,x) = 0
+       DO y = 1, N
+          a(y,x) = x + y
+          b(y+1,x) = 0 
+       END DO
+    END DO
+
+    !$omp target parallel num_threads(OMPVV_NUM_THREADS_DEVICE) map(tofrom: a, b)
+    !$omp loop collapse(1)
+    DO x = 1, N
+       DO y = 1, N
+          b(y+1,x) = b(y,x) + a(y,x)
+       END DO
+    END DO
+    !$omp end loop
+    !$omp end target parallel
+
+    DO x = 1, N
+       temp_total = 0
+       DO y = 1, N+1
+          OMPVV_TEST_AND_SET(errors, (temp_total - b(y,x)) .ne. 0)
+       END DO
+       IF ( y .ne. N+1 ) THEN
+          temp_total = temp_total + a(y,x)
```

Comment:
This leads to out of bounds: The loop goes from `y = 1` to `N+1` – and after the loop, `y = N + 2`. However, `a(1:N)` and thus, `y == 130` is out of bounds (`N = 128`).

In the second function, `test_collapse2`, it is inside the DO loop, which avoids the out of bounds.

Still: It is unclear to me what that `temp_total` is good for – it is not used for any error checking, but it also does not show up in an OpenMP context. Additionally, it is not initialized. (Those issues also apply to the second function.)

Reply:
@tob2 thank you for finding this error; the IF-statement should be moved into the DO-loop as you suggested.
`temp_total` is used to check whether the innermost loop that is out of the scope of the collapsed loops was parallelized or not by checking possible race condition (race condition may occur to `b(y,x)` if the innermost loop was parallelized.).

Reply:
The code was changed to properly initialize `temp_total`.